singlepoll/linux.go

```diff
@@ -0,0 +1,194 @@
+// +build linux
+
+package singlepoll
+
+import (
+	"context"
+	"errors"
+	"runtime"
+	"sync"
+	"syscall"
+
+	"github.com/gxed/eventfd"
+	logging "github.com/ipfs/go-log"
+)
+
+var (
+	ErrUnsupportedMode error = errors.New("only 'w' and 'r' modes are supported on this arch")
+)
+
+var (
+	initOnce sync.Once
+	workChan chan interface{}
+	log      logging.EventLogger = logging.Logger("reuseport-poll")
+)
+
+type addPoll struct {
+	fd     int
+	events uint32
+	ctx    context.Context
+	wakeUp chan<- error
+}
+
+type ctxDone struct {
+	fd int
+}
+
+func PollPark(reqctx context.Context, fd int, mode string) error {
+	initOnce.Do(func() {
+		workChan = make(chan interface{}, 128)
+		go worker()
+	})
+
+	events := uint32(0)
+	for _, c := range mode {
+		switch c {
+		case 'w':
+			events |= syscall.EPOLLOUT
+		case 'r':
+			events |= syscall.EPOLLIN
+		default:
+			return ErrUnsupportedMode
+		}
+
+	}
+
+	wakeUp := make(chan error)
+	workChan <- addPoll{
+		fd:     fd,
+		events: events,
+		ctx:    reqctx,
+		wakeUp: wakeUp,
+	}
+
+	return <-wakeUp
```

Comment:
Should we consider selecting over this response and the context that was passed in? I guess if the context is cancelled in that case we might not get the right error?